packages/integration-sdk-runtime/src/synchronization/batchBySize.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  Entity,
+  IntegrationLogger,
+  Relationship,
+} from '@jupiterone/integration-sdk-core';
+import { UploadDataLookup } from '.';
+import { MAX_BATCH_SIZE } from './shrinkBatchRawData';
+const BATCH_THRESHOLD = 0.9; //will stop chunking if chunk size is between BATCH_THRESHOLD*sizeInBytes and sizeInBytes
+
+export function chunkBySize<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  sizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  if (sizeInBytes > MAX_BATCH_SIZE) {
+    logger.error({}, 'batch size is too big');
+    throw new Error('batch size is too big');
+  }
+  return chunk(data, sizeInBytes, logger);
+}
+export function chunk<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  sizeInBytes: number,
+  logger: IntegrationLogger,
+): T[K][][] {
+  const chunkedData: T[K][][] = [];
+  if (getSizeOfObject(data) < sizeInBytes) {
+    return [data]; //Just one chunk of all data
+  }
+  let bestIndex = binarySearch(data, sizeInBytes);
+  if (bestIndex <= 0) {
+    //If the first entity is too big
+    handleBinarySearchError(data, sizeInBytes, logger); // we remove rawdata
+    bestIndex = 1; //and send it to the persister alone. TODO: find a way to avoid sending this entities as a single call. It shouldn't be an issue since this only happens if there is a single entity above sizeInBytes
+  }
+  chunkedData.push(data.slice(0, bestIndex));
+  if (bestIndex !== data.length)
+    chunkedData.push(
+      ...chunk(data.slice(bestIndex), sizeInBytes, logger), //chunk the rest of the data
+    );
+  return chunkedData;
+}
+//If the batch doesnt fit, we try with half, and then half of half ... until it fits.
+//We repeat until we reach the Threshold or until the while condition.
+//We are binary-searching for the best spot to chunk
+function binarySearch<T extends UploadDataLookup, K extends keyof T>(
+  data: T[K][],
+  targetSize: number,
+): number {
+  let left: number = 0;
+  let right: number = data.length - 1;
+  let bestIndex: number = -1;
+  const thSize = targetSize * BATCH_THRESHOLD;
+  let mid: number = Math.floor((left + right) / 2);
+  let size = getSizeOfObject(data.slice(0, mid));
+  while (left <= right) {
+    if (size <= targetSize) {
+      bestIndex = mid;
+      if (size >= thSize) {
+        return mid; //if its **close enough** to the target size
+      }
+    }
+    if (targetSize < size) {
+      right = mid - 1;
+      if (left <= right) {
+        const newMid = Math.floor((left + right) / 2);
+        size = size - getSizeOfObject(data.slice(newMid, mid));
```

Comment:
I think `JSON.stringify()` is probably the most expensive operation we're using here, so the most efficient algorithm would only calculate the size of each object once. It seems like the binary search algorithm here might be calculating the size of each given object more than once, in the event that `if (targetSize < size)` ever evaluates to `true`.

What about looping through each object, calculating its size just once, and then building up your chunks one-by-one?

Reply:
Maybe something like this: 

```ts
export function batchGraphObjectsBySizeInBytes<T extends UploadDataLookup, K extends keyof T>(
  graphObjects: T[K][],
  maximumBatchSizeInBytes: number,
  logger: IntegrationLogger,
): T[K][][] {
  const batches: T[K][][] = [];

  let currentBatch: T[K][] = [];
  let currentBatchSize: number = 0;
  for (const graphObject of graphObjects) {
    let graphObjectSizeInBytes = getSizeOfObject(graphObject);

    if (graphObjectSizeInBytes > maximumBatchSizeInBytes) {
      const rawDataSizeInBytes = getSizeOfObject(graphObject._rawData);

      if (graphObjectSizeInBytes-rawDataSizeInBytes > maximumBatchSizeInBytes) {
        logger.warn({ 
          _key, _type, maximumBatchSizeInBytes, graphObjectSizeInBytes, rawDataSizeInBytes,
         }, 'Graph object is larger than batch size in bytes; cannot upload graph object.');
         continue;
      } else {
        logger.info({ 
          _key, _type, maximumBatchSizeInBytes, graphObjectSizeInBytes, rawDataSizeInBytes,
         }, 'Graph object is larger than batch size in bytes; removing raw data.');
         graphObjectSizeInBytes -= rawDataSizeInBytes;
         graphObject._rawData = undefined;
      }
    }

    if (graphObjectSizeInBytes + currentBatchSize >= maximumBatchSizeInBytes) {
      batches.push(currentBatch);
      currentBatch = [];
      currentBatchSize = 0;
    }
    currentBatch.push(graphObject);
    currentBatchSize += graphObjectSizeInBytes;
  }

  if (currentBatch.length > 0) {
    batches.push(currentBatch);
  }
  return batches;
}
```